### PR TITLE
Add function which normalises input cubes according to a reference

### PR DESCRIPTION
**Pull request by brhooper**
Addresses https://github.com/metoppv/mo-blue-team/issues/520
Acceptance test data: https://github.com/metoppv/improver_test_data/pull/20

Description
Adds plugin and CLI which takes a cubelist and a reference cube as inputs and adjusts the data in the input cubelist so that the sum total of this data is equal to the data in reference cube. Also adds unit tests for the new function.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Review by Katie-Howard — CHANGES_REQUESTED** [review]
I've looked through all these changes, and they seem generally fine to me. There are just a few minor comments that I have made on the code. It all seems to work fine, and the new unit tests/acceptance tests also run through without any problems. I've had a little look at the new acceptance test data and that too looks ok 👍

File: improver/cli/enforce_consistent_forecasts.py [hunk 1]
```diff
@@ -95,8 +95,10 @@ def process(
     """
     from iris.cube import CubeList
 
-    from improver.utilities.enforce_consistency import EnforceConsistentForecasts
     from improver.utilities.flatten import flatten
+    from improver.utilities.forecast_reference_enforcement import (
+        EnforceConsistentForecasts,
```
[the remark this answers is not in the document]
Comment (brhooper) [reply improver/cli/enforce_consistent_forecasts.py:100]: I removed this comma, then when I ran black to check the formatting it added it back in, so I think the comma will have to stay

File: improver/cli/normalise_to_reference.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+@cli.clizefy
+@cli.with_output
+def process(
+    cubes: cli.inputcubelist, reference: cli.inputcube, ignore_zero_total: bool = False
+):
+    """Module to enforce that the sum of data in input cubes is equal to the
+    corresponding data in reference.
+
+    The data are updated as follows, if cubes contains 2 cubes with data points a and b
```
[the remark this answers is not in the document]
Comment (brhooper) [reply improver/cli/normalise_to_reference.py]: I've edited this docstring to now read

"The data are updated as follows, if input cubes is a cubelist of 2 cubes with data points a and b respectively and corresponding reference r, then:"

hopefully this makes it a bit easier to understand what I mean, but if you'd like me to change it again or have a suggestion then I'm happy to take another look.

File: improver/cli/normalise_to_reference.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    respectively and corresponding reference r, then:
+
+    a_new = r * a / (a + b)
```
[the remark this answers is not in the document]
Comment (brhooper) [reply improver/cli/normalise_to_reference.py]: I've added brackets as suggested

File: improver/cli/enforce_consistent_forecasts.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Katie-Howard) [reply improver/cli/enforce_consistent_forecasts.py:100]: That's fine.

**Review by Katie-Howard — DISMISSED** [review]
I've looked through the changes and I'm happy with them now, thanks 👍

File: improver/utilities/forecast_reference_enforcement.py [hunk 4]
```diff
@@ -346,3 +346,34 @@ def normalise_to_reference(
...
+            a single name or a list of names.
+
+    Returns:
+        - A cubelist containing all cubes with names which match cube_names
+        - A cubelist containing all cubes with names which do not match cube_names
+    """
+
+    desired_cubes = iris.cube.CubeList()
+    other_cubes = iris.cube.CubeList()
+
```
Comment (bayliffe) [inline improver/utilities/forecast_reference_enforcement.py:369]: I wonder whether it is tidier to make this function return all the cubes as `desired_cubes` if `cube_names == None`. It would remove the logic in the CLI in relation to return_name, and allow this function to be called in other places where the status of return_name is similarly unknown.

File: improver/cli/normalise_to_reference.py [hunk 5]
```diff
@@ -37,37 +37,70 @@
...
     """
 
-    from improver.utilities.forecast_reference_enforcement import normalise_to_reference
+    from improver.utilities.forecast_reference_enforcement import (
+        normalise_to_reference,
+        split_cubes_by_name,
+    )
 
-    return normalise_to_reference(cubes, reference, ignore_zero_total)
+    cubes = cubes[0]
```
Comment (bayliffe) [inline improver/cli/normalise_to_reference.py:85]: This should not be here.

**Review by bayliffe — CHANGES_REQUESTED** [review]
A couple of comments and now I can't read anymore (migraine aura's are trippy). Bailing.

File: improver/cli/normalise_to_reference.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+
+@cli.clizefy
+@cli.with_output
+def process(
+    *cubes: cli.inputcubelist,
+    reference_name: cli.comma_separated_list,
+    return_name: cli.comma_separated_list = None,
+    ignore_zero_total: bool = False,
+):
```
Comment (MoseleyS) [inline improver/cli/normalise_to_reference.py:44]:
```suggestion
) -> Cube:
```

File: improver/cli/normalise_to_reference.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            corresponding value in reference is non-zero. If True this error will not be
+            raised - instead the data in the updated cubes will remain as zero values,
+            if False an error will be raised.
+
+    Returns:
+        iris.cube.Cube or iris.cube.CubeList:
+            A cubelist containing cubes with identical metadata to those in input cubes,
+            but with data updated so that the total of the data in these cubes is equal
+            to the reference. If only one cube is being returned, then a cube is
+            returned instead of a cubelist.
```
Comment (MoseleyS) [inline improver/cli/normalise_to_reference.py:76]: As discussed, we will only support returning one Cube, not a list of Cubes, at least until we decide otherwise.

File: improver/cli/normalise_to_reference.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        msg = (
+            f"Exactly one cube with a name matching reference_name is required, but "
+            f"{len(reference_cube)} were found."
+        )
+        raise ValueError(msg)
+
+    output = normalise_to_reference(input_cubes, reference_cube, ignore_zero_total)
+
+    if return_name is not None:
+        output, _ = split_cubes_by_name(output, return_name)
```
Comment (MoseleyS) [inline improver/cli/normalise_to_reference.py:101]:
```suggestion
        output, = output.extract(return_name)
```

File: improver_tests/utilities/test_split_cubes_by_name.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+"""Unit tests for the "forecast_reference_enforcement.split_cubes_by_name" function."""
+import iris
+import numpy as np
+import pytest
+
+from improver.synthetic_data.set_up_test_cubes import set_up_percentile_cube
+from improver.utilities.forecast_reference_enforcement import split_cubes_by_name
+
+
+def create_test_cubes(n_cubes, name):
```
Comment (MoseleyS) [inline improver_tests/utilities/test_split_cubes_by_name.py:40]: Please add type-hinting and a doc-string. Also, you could just use `set_up_variable_cube`, no need for percentiles for these tests, but I don't mind if this is left as it is.

File: improver_tests/utilities/test_normalise_to_reference.py [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+        output_sum = output[0].data + output[1].data + output[2].data
+
+        reference_with_zeroes = reference_percentile_cube.copy()
+        reference_with_zeroes.data[0, :, :] = 0.0
+
+        assert output == expected_percentile_cubes
+        assert np.array_equal(output_sum, reference_with_zeroes.data)
+
+
+def test_single_input_cube(input_percentile_cubes, reference_percentile_cube):
```
Comment (MoseleyS) [inline improver_tests/utilities/test_normalise_to_reference.py:259]: I'm not sure why we would ever do this. Perhaps it should raise an exception instead?
Also - add a test for the inputs being an empty cube list.

File: improver/cli/normalise_to_reference.py (hunk 5 above)
Comment (bayliffe) [reply improver/cli/normalise_to_reference.py:85]: I think I understand why this was put here, but it's not doing what you want when called from the command line; actually the behaviour of the acceptance tests is rather weird. From the command line, calling this CLI with multiple cubes, the cubes variable becomes something like this:

```
(
    [
        <iris 'Cube' of probability_of_lwe_thickness_of_precipitation_amount_above_threshold / (1) (lwe_thickness_of_precipitation_amount: 27; projection_y_coordinate: 970; projection_x_coordinate: 1042)>
    ],
    [
        <iris 'Cube' of probability_of_thickness_of_rainfall_amount_above_threshold / (1) (thickness_of_rainfall_amount: 27; projection_y_coordinate: 970; projection_x_coordinate: 1042)>
    ],
    [
        <iris 'Cube' of probability_of_lwe_thickness_of_sleetfall_amount_above_threshold / (1) (lwe_thickness_of_sleetfall_amount: 27; projection_y_coordinate: 970; projection_x_coordinate: 1042)>
    ],
    [
        <iris 'Cube' of probability_of_lwe_thickness_of_snowfall_amount_above_threshold / (1) (lwe_thickness_of_snowfall_amount: 27; projection_y_coordinate: 970; projection_x_coordinate: 1042)>
    ]
)
```

Meaning that your cubes = cubes[0] line just gives you:
```
cubes = [
    <iris 'Cube' of probability_of_lwe_thickness_of_precipitation_amount_above_threshold / (1) (lwe_thickness_of_precipitation_amount: 27; projection_y_coordinate: 970; projection_x_coordinate: 1042)>
],
```

However, your acceptance tests create a single list wrapped in a tuple:

```
(
    [
        <iris 'Cube' of probability_of_lwe_thickness_of_precipitation_amount_above_threshold / (1) (lwe_thickness_of_precipitation_amount: 27; projection_y_coordinate: 970; projection_x_coordinate: 1042)>,
        <iris 'Cube' of probability_of_thickness_of_rainfall_amount_above_threshold / (1) (thickness_of_rainfall_amount: 27; projection_y_coordinate: 970; projection_x_coordinate: 1042)>,
        <iris 'Cube' of probability_of_lwe_thickness_of_sleetfall_amount_above_threshold / (1) (lwe_thickness_of_sleetfall_amount: 27; projection_y_coordinate: 970; projection_x_coordinate: 1042)>,
        <iris 'Cube' of probability_of_lwe_thickness_of_snowfall_amount_above_threshold / (1) (lwe_thickness_of_snowfall_amount: 27; projection_y_coordinate: 970; projection_x_coordinate: 1042)>
    ]
)
```
Hence your command yields a simple list that can be iterated over.
I have absolutely no idea why calling the CLI from the acceptance test is behaving differently to the command line and it perturbs me. However, the solution is to use the `flatten` utility which is used in some other CLIs that accept a cubelist. This will enable the acceptance tests and command line calls to behave in the same way.

```
from improver.utilities.flatten import flatten

cubes = flatten(cubes)
```

File: improver/utilities/forecast_reference_enforcement.py (hunk 4 above)
Comment (brhooper) [reply improver/utilities/forecast_reference_enforcement.py:377]: Done.

File: improver/cli/normalise_to_reference.py [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+from improver import cli
+
+
+@cli.clizefy
+@cli.with_output
+def process(
+    *cubes: cli.inputcubelist,
+    reference_name: cli.comma_separated_list,
+    return_name: cli.comma_separated_list = None,
```
[the remark this answers is not in the document]
Comment (brhooper) [reply improver/cli/normalise_to_reference.py]: Done

File: improver/cli/normalise_to_reference.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    a_new = r * (a / (a + b))
+    b_new = r * (b / (a + b))
+
+    which ensures that r = a_new + b_new.
+
+    Args:
+        cubes (List of iris.cube.Cube): A list of cubes containing both the cubes to be
+            updated and the reference cube. The reference cube will be identified by
+            matching the cube name to reference_name.
+        reference_name (str): The name of the reference cube, this should match exactly
```
[the remark this answers is not in the document]
Comment (brhooper) [reply improver/cli/normalise_to_reference.py]: Done

File: improver/utilities/forecast_reference_enforcement.py [hunk 13]
```diff
@@ -242,3 +243,137 @@ def process(self, forecast: Cube, reference_forecast: Cube) -> Cube:
...
+        ValueError: If there are instances where the total of the input cubes is zero
+            but the corresponding value in reference is non-zero. This error can be
+            ignored if ignore_zero_total is true.
+
+    Returns:
+        Cubelist with length equal to the length of cubes. Each cube in the returned
+        cubelist will have metadata matching the cube in the same position in input
+        cubes, but containing different data.
+    """
+    input = cubes.copy()
```
[the remark this answers is not in the document]
Comment (brhooper) [reply improver/utilities/forecast_reference_enforcement.py]: Done

File: improver/utilities/forecast_reference_enforcement.py [hunk 14]
```diff
@@ -242,3 +243,137 @@ def process(self, forecast: Cube, reference_forecast: Cube) -> Cube:
...
+
+    if coord_mismatch and not is_probability(reference):
+        msg = (
+            f"The dimension coordinates on the input cubes and the reference did not "
+            f"all match. The following coordinates were found to differ: "
+            f"{mismatching_coords}."
+        )
+        raise ValueError(msg)
+
+    total = input[0].copy().data
```
[the remark this answers is not in the document]
Comment (brhooper) [reply improver/utilities/forecast_reference_enforcement.py]: Done

File: improver/cli/normalise_to_reference.py (hunk 6 above)
Comment (brhooper) [reply improver/cli/normalise_to_reference.py:44]: I don't think I can add this, as the acceptance tests complain about importing numpy (presumably via iris) if I do

File: improver/cli/normalise_to_reference.py (hunk 7 above)
Comment (brhooper) [reply improver/cli/normalise_to_reference.py]: Changed

File: improver/cli/normalise_to_reference.py (hunk 8 above)
Comment (brhooper) [reply improver/cli/normalise_to_reference.py]: I've changed this to use `output.extract_cube(return_name)`

File: improver_tests/acceptance/test_normalise_to_reference.py [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+        run_cli(args)
+        acc.compare(output_path, kgo_path)
+    else:
+        args = [
+            inputs,
+            "--reference-name",
+            reference_name,
+            "--output",
+            output_path,
+        ]
```
[the remark this answers is not in the document]
Comment (brhooper) [reply improver_tests/acceptance/test_normalise_to_reference.py]: Change made. I've also adapted this test to include a test using probability data, following @bayliffe's suggestion

File: improver_tests/utilities/test_split_cubes_by_name.py [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    if n_cubes != 0:
+        cube = set_up_percentile_cube(data, percentiles, name=name, units="m s-1")
+        for index in range(n_cubes):
+            output.append(cube.copy())
+
+    return output
+
+
+@pytest.mark.parametrize("n_match, n_other", ((0, 1), (1, 0), (1, 1), (2, 1), (1, 2)))
+def test_split_cubes_by_name(n_match, n_other):
```
[the remark this answers is not in the document]
Comment (brhooper) [reply improver_tests/utilities/test_split_cubes_by_name.py]: Done

File: improver_tests/utilities/test_split_cubes_by_name.py (hunk 9 above)
Comment (brhooper) [reply improver_tests/utilities/test_split_cubes_by_name.py]: Done

File: improver_tests/utilities/test_normalise_to_reference.py (hunk 10 above)
Comment (brhooper) [reply improver_tests/utilities/test_normalise_to_reference.py]: Added an exception for when `len(cubes) < 2' and modified this test to look for that exception instead

File: improver_tests/utilities/test_normalise_to_reference.py [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+
+from improver.synthetic_data.set_up_test_cubes import (
+    set_up_percentile_cube,
+    set_up_probability_cube,
+)
+from improver.utilities.forecast_reference_enforcement import normalise_to_reference
+
+
+@pytest.fixture
+def shape():
```
[the remark this answers is not in the document]
Comment (brhooper) [reply improver_tests/utilities/test_normalise_to_reference.py]: Agreed, done

File: improver/cli/normalise_to_reference.py (hunk 6 above)
Comment (MoseleyS) [reply improver/cli/normalise_to_reference.py:44]: Oh yes, I think that's right, so leave this out.

File: improver/utilities/forecast_reference_enforcement.py [hunk 18]
```diff
@@ -242,3 +243,137 @@ def process(self, forecast: Cube, reference_forecast: Cube) -> Cube:
...
+    Returns:
+        Cubelist with length equal to the length of cubes. Each cube in the returned
+        cubelist will have metadata matching the cube in the same position in input
+        cubes, but containing different data.
+    """
+    input = cubes.copy()
+
+    # check cube compatibility
+    reference_dim_coords = reference.coords(dim_coords=True)
+    n_dims_mismatch = False
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/utilities/forecast_reference_enforcement.py:289]: Agreed.

File: improver_tests/acceptance/test_normalise_to_reference.py [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+from . import acceptance as acc
+
+pytestmark = [pytest.mark.acc, acc.skip_if_kgo_missing]
+CLI = acc.cli_name_with_dashes(__file__)
+run_cli = acc.run_cli(CLI)
+
+
+@pytest.mark.parametrize(
+    "forecast_type,ignore_zero_total",
+    (("probability", False), ("percentile", False), ("percentile", True)),
```
Comment (MoseleyS) [inline improver_tests/acceptance/test_normalise_to_reference.py:44]: As `("probability", True)` isn't included, I don't think the `probability/kgo.nc` is used.

**Review by MoseleyS — CHANGES_REQUESTED** [review]
Just one more thing...

File: improver_tests/acceptance/test_normalise_to_reference.py (hunk 19 above)
Comment (brhooper) [reply improver_tests/acceptance/test_normalise_to_reference.py:44]: The probability inputs don't contain any instances where the total of the inputs is zero but the reference is non-zero. Due to this the CLI can happily run with `ignore_zero_total=False` (the default), so I don't think that a change is required here.

Comment (MoseleyS) [reply improver_tests/acceptance/test_normalise_to_reference.py:44]: Sorry - I had misread the `if..elif..else` block

**Review by MoseleyS — APPROVED** [review]

**Review by Katie-Howard — APPROVED** [review]
These changes look good to me.

**Review by bayliffe — APPROVED** [review]
I've not reviewed the unit tests, but you seem to have ample reviewers, so I think that's okay. I'm happy with this, thanks Ben.

Merging, along with the acceptance test data.
